File: src/Services/Tasks/Libr4.Tasks.Domain/TimeTracking/TimeTrackingErrors.cs
using Libr4.Shared.Kernel.Errors;
using Libr4.Shared.Kernel.Results;

namespace Libr4.Tasks.Domain.TimeTracking;

public static class TimeTrackingErrors
{
    public static readonly Error SessionNotFound = Error.NotFound("session.not_found", "Time session not found");
    public static readonly Error NotSessionOwner = Error.Forbidden("session.not_owner", "You are not the owner of this session");
    public static readonly Error SessionNotActive = Error.Conflict("session.not_active", "Session is not active");
    public static readonly Error SessionAlreadyStopped = Error.Conflict("session.already_stopped", "Session is already stopped");
    public static readonly Error InvalidSessionStatus = Error.Conflict("session.invalid_status", "Invalid session status transition");
    public static readonly Error ScreenshotNotFound = Error.NotFound("screenshot.not_found", "Screenshot not found");
    public static readonly Error ActivityLogNotFound = Error.NotFound("log.not_found", "Activity log not found");
    public static readonly Error AlertNotFound = Error.NotFound("alert.not_found", "Anti-cheat alert not found");
    public static readonly Error ReportNotFound = Error.NotFound("report.not_found", "Time report not found");
    public static readonly Error SettingsNotFound = Error.NotFound("settings.not_found", "Time tracking settings not found");
}

File: src/Services/Tasks/Libr4.Tasks.Domain.TimeTracking.FSharp/TimeTrackingErrors.fs
namespace Libr4.Tasks.Domain.TimeTracking.FSharp

/// Domain errors for time tracking operations
module TimeTrackingErrors =

    /// Error type for time tracking domain
    type TimeTrackingError =
        | SessionNotFound
        | NotSessionOwner
        | SessionNotActive
        | SessionAlreadyStopped
        | InvalidSessionStatus
        | ScreenshotNotFound
        | ActivityLogNotFound
        | AlertNotFound
        | ReportNotFound
        | SettingsNotFound
        | InvalidDuration
        | InvalidRate
        | InvalidActivityLevel

    /// Convert error to message
    let errorMessage = function
        | SessionNotFound -> "Time session not found"
        | NotSessionOwner -> "You are not the owner of this session"
        | SessionNotActive -> "Session is not active"
        | SessionAlreadyStopped -> "Session is already stopped"
        | InvalidSessionStatus -> "Invalid session status transition"
        | ScreenshotNotFound -> "Screenshot not found"
        | ActivityLogNotFound -> "Activity log not found"
        | AlertNotFound -> "Anti-cheat alert not found"
        | ReportNotFound -> "Time report not found"
        | SettingsNotFound -> "Time tracking settings not found"
        | InvalidDuration -> "Invalid session duration"
        | InvalidRate -> "Invalid hourly rate"
        | InvalidActivityLevel -> "Invalid activity level"

    /// Validation result type
    type ValidationResult<'T> = Result<'T, TimeTrackingError>

    /// Validate session ownership
    let validateSessionOwner (userId: System.Guid) (session: TimeSessionRecord) : ValidationResult<unit> =
        if session.userId = userId then Ok ()
        else Error NotSessionOwner

    /// Validate session is active
    let validateSessionActive (session: TimeSessionRecord) : ValidationResult<unit> =
        match session.status with
        | SessionStatus.Active -> Ok ()
        | _ -> Error SessionNotActive

    /// Validate session not stopped
    let validateSessionNotStopped (session: TimeSessionRecord) : ValidationResult<unit> =
        match session.status with
        | SessionStatus.Completed | SessionStatus.Abandoned -> Error SessionAlreadyStopped
        | _ -> Ok ()

    /// Validate duration
    let validateDuration (minutes: float) : ValidationResult<float> =
        if minutes > 0.0 && minutes <= 1440.0 then Ok minutes
        else Error InvalidDuration

    /// Validate hourly rate
    let validateRate (rate: decimal option) : ValidationResult<decimal option> =
        match rate with
        | None -> Ok None
        | Some r when r > 0m && r <= 10000m -> Ok (Some r)
        | _ -> Error InvalidRate

    /// Validate activity level
    let validateActivityLevel (level: float option) : ValidationResult<float option> =
        match level with
        | None -> Ok None
        | Some l when l >= 0.0 && l <= 100.0 -> Ok (Some l)
        | _ -> Error InvalidActivityLevel
